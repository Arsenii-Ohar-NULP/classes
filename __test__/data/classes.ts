import Class from 'components/classes/Class';

export const sampleFiveClasses: Class[] = [
    {
        id: 1,
        title: 'Class 1',
        description: 'This is class 1',
        teacher_id: 1,
        students_number: 1
    },
    {
        id: 2,
        title: 'Class 2',
        description: 'This is class 2',
        teacher_id: 2,
        students_number: 1
    },
    {
        id: 3,
        title: 'Class 3',
        description: 'This is class 3',
        teacher_id: 3,
        students_number: 1

    },
    {
        id: 4,
        title: 'Class 4',
        description: 'This is class 4',
        teacher_id: 4,
        students_number: 1
    },
    {
        id: 5,
        title: 'Class 5',
        description: 'This is class 5',
        teacher_id: 5,
        students_number: 1
    }
]