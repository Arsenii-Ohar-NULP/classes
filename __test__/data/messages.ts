import Message from 'pages/class/Message';

const sampleMessages : Message[] = [
    {
        id: 1,
        content: 'First',
        date: 'Yesterday',
        user: 1,
        cls: 1,
        fullname: 'Arsenii Ohar',
        username: 'arseniyohar'
    },
    {
        id: 2,
        content: 'First',
        date: 'Yesterday',
        user: 1,
        cls: 1,
        fullname: 'Arsenii Ohar',
        username: 'arseniyohar'
    },
    {
        id: 3,
        content: 'First',
        date: 'Yesterday',
        user: 1,
        cls: 1,
        fullname: 'Arsenii Ohar',
        username: 'arseniyohar'
    }

]