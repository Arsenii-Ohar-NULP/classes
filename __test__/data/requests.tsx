import { JoinRequest } from 'pages/class/JoinRequest'

export const getRequestsWithClassId = (classId: number, number: number) => {
    const requests: JoinRequest[] = []
    for (let i = 0; i < number; i++){
        requests.push(
            {
                userId: i + 1,
                classId: classId
            }
        )
    }

    return requests;
}