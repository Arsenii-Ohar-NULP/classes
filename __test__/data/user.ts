import User, { Role } from 'components/account/User';

export const sampleUser: User = {
    id: 1,
    username: 'seniorohar',
    firstName: 'Arsenii',
    lastName: 'Ohar',
    email: 'arsen.ogar@gmail.com',
    phone: '+380987669293',
    role: Role.Teacher   
}

export const sampleFiveUsers: User[] = [
    {
        id: 1,
        username: 'seniorohar',
        firstName: 'Arsenii',
        lastName: 'Ohar',
        email: 'arsen.ogar@gmail.com',
        phone: '+380987669293',
        role: Role.Student
    },
    {
        id: 2,
    username: 'vitalikpikh',
    firstName: 'Vitalik',
    lastName: 'Pikhotskiy',
    email: 'vitaliipik@gmail.com',
    phone: '+380967629253',
    role: Role.Student
    },
    {
        id: 3,
    username: 'svyatoslavshainoha',
    firstName: 'Svyatoslav',
    lastName: 'Shainoha',
    email: 'vitaliipik@gmail.com',
    phone: '+380967629253',
    role: Role.Student
    },
      {
        id: 4,
    username: 'olehrysin',
    firstName: 'Oleh',
    lastName: 'Rysin',
    email: 'olehrysin@gmail.com',
    phone: '+380967323273',
    role: Role.Student
    },
    {
        id: 5,
    username: 'johnadams',
    firstName: 'John',
    lastName: 'Adams',
    email: 'johnadams@gmail.com',
    phone: '+380967723273',
    role: Role.Student
    }
]