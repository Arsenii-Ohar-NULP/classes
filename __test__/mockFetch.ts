export const mockFetch = () => {
    global.fetch = jest.fn();
}