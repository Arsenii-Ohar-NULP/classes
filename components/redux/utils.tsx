import {getAccessToken} from "../account/TokenService";

export const CLASSES_API_URL = `${process.env["NEXT_PUBLIC_API_URL"]}/api/v1`;


export function authHeaders(headers: Headers) {
    const token = getAccessToken()
    if (token) {
        headers.set("authorization", `Bearer ${token}`)
    }

    return headers;
}

