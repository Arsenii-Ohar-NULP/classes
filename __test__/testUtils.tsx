import React, { PropsWithChildren } from 'react';
import { render, renderHook } from '@testing-library/react';
import type { RenderOptions } from '@testing-library/react';
import { configureStore } from '@reduxjs/toolkit';
import type { PreloadedState } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';

import type { RootState, AppStore } from 'pages/redux/store';
// As a basic setup, import your same slice reducers
import { authInitialState, authSlice } from 'pages/redux/auth';
import { classSlice, classesInitialState } from 'pages/redux/classes';

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<RootState>;
  store?: AppStore;
}

export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = { auth: authInitialState, classes: classesInitialState },
    // Automatically create a store instance if no store was passed in
    store = configureStore({
      reducer: { auth: authSlice.reducer, classes: classSlice.reducer },
      preloadedState,
    }),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  function Wrapper({ children }: PropsWithChildren<object>): JSX.Element {
    return <Provider store={store}>{children}</Provider>;
  }

  // Return an object with the store and all of RTL's query functions
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

export function renderHookWithProviders(
  ui,
  {
    preloadedState = { auth: authInitialState, classes: classesInitialState },
    // Automatically create a store instance if no store was passed in
    store = configureStore({
      reducer: { auth: authSlice.reducer, classes: classSlice.reducer },
      preloadedState,
    }),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  function Wrapper({ children }: PropsWithChildren<object>): JSX.Element {
    return <Provider store={store}>{children}</Provider>;
  }

  // Return an object with the store and all of RTL's query functions
  return { store, ...renderHook(() => ui(), { wrapper: Wrapper, ...renderOptions }) };
}
