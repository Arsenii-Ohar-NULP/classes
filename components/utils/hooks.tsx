import { useEffect, useState } from 'react';
import { useRouter } from 'next/router';
import { AuthStatus, authActions } from 'components/redux/auth';
import { useAppSelector, useAppDispatch } from 'components/redux/store';
import User from 'components/account/User';
import InvalidCredentials from 'components/errors/InvalidCredentials';
import { getUserInfo } from 'components/account/UserService';
import { getAccessToken, removeToken } from 'components/login/AuthService';
import { socket } from './socket';

function useLoginRedirect() {
  const router = useRouter();
  const authStatus = useAppSelector((state) => state.auth.status);
  const dispatch = useAppDispatch();

  useEffect(() => {
    if (!getAccessToken()) {
      if (authStatus !== AuthStatus.LOGGED_OUT) {
        dispatch(authActions.logout());
      }

      router.push('/login');
    }
  }, [authStatus]);
}

function useUserData() {
  const router = useRouter();
  const [user, setUser] = useState<User>(null);
  const dispatch = useAppDispatch();

  useEffect(() => {
    if (!user) {
      const token = getAccessToken();
      getUserInfo(token)
        .then((fetchedUser) => setUser(fetchedUser))
        .catch((e) => {
          if (e instanceof InvalidCredentials) {
            removeToken();
            dispatch(authActions.logout());
            router.push('/login');
          }
        });
    }
  });

  return user;
}

function useMainPageRedirect() {
  const mainPagePath = '/classes';
  const router = useRouter();
  const authStatus = useAppSelector((state) => state.auth.status);
  const dispatch = useAppDispatch();
  useEffect(() => {
    const token = getAccessToken();
    if (token) {
      getUserInfo(token)
      .then((user) => {
        if (authStatus === AuthStatus.LOGGED_OUT){
          dispatch(authActions.login(user))
        }
        router.push(mainPagePath)
      })
      .catch(error => {
        if (error instanceof InvalidCredentials){
          return;
        }
      })
    }
  }, [authStatus]);
}

function useMessageReplacer(replacers: object) {
  function normalize(message: string) {
    for (const replacer of Object.keys(replacers)) {
      message = message.replace(
        replacer.toString(),
        replacers[replacer.toString()].toString()
      );
    }

    return message;
  }

  return normalize;
}

function useBootstrap() {
  useEffect(() => {
    require('bootstrap/dist/js/bootstrap.bundle.js');
  });
}

function useSocket(){
  useEffect(() => {
    socket.connect();

    return () => {
      socket.disconnect();
    }
  }, [])
}

export {
  useLoginRedirect,
  useUserData,
  useMainPageRedirect,
  useMessageReplacer,
  useBootstrap,
  useSocket
};
