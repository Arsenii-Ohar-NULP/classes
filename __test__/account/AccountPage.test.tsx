import { sampleUser } from '__test__/data/user';
import { renderWithProviders } from '__test__/testUtils';
import AccountPage from 'app/main/account/page';
import { useUserData } from 'components/utils/hooks';
import React from 'react';

jest.mock('components/utils/hooks');

describe('account page test', () => {
  it('should match a snapshot', () => {
    jest.mocked(useUserData).mockReturnValueOnce(sampleUser);
    const page = renderWithProviders(<AccountPage />);
    expect(page.container).toMatchSnapshot();
  });

  it('given no user auth, should match a snapshot', () => {
    jest.mocked(useUserData).mockReturnValueOnce(null);
    const page = renderWithProviders(<AccountPage />);
    expect(page.container).toMatchSnapshot();
  });
});
