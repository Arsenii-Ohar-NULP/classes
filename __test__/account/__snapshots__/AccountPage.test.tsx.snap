// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`account page test given no user auth, should match a snapshot 1`] = `<div />`;

exports[`account page test should match a snapshot 1`] = `
<div>
  <div
    class="container"
  >
    <h1
      class="p-2 text-center"
    >
      Personal Information
    </h1>
    <div
      class="d-flex justify-content-center"
    >
      <div
        class="d-inline-block"
      >
        <div
          class="d-flex flex-column gap"
        >
          <div
            class="d-flex gap-5"
          >
            <h6
              class="col"
            >
              Username
            </h6>
            <h6
              class="col"
            >
              seniorohar
            </h6>
          </div>
          <hr />
          <div
            class="d-flex gap-5"
          >
            <h6
              class="col"
            >
              Email
            </h6>
            <h6
              class="col"
            >
              arsen.ogar@gmail.com
            </h6>
          </div>
          <hr />
          <div
            class="d-flex gap-5"
          >
            <h6
              class="col"
            >
              Phone
            </h6>
            <h6
              class="col"
            >
              +380987669293
            </h6>
          </div>
          <hr />
          <div
            class="d-flex gap-5"
          >
            <h6
              class="col"
            >
              First Name
            </h6>
            <h6
              class="col"
            >
              Arsenii
            </h6>
          </div>
          <hr />
          <div
            class="d-flex gap-5"
          >
            <h6
              class="col"
            >
              Last Name
            </h6>
            <h6
              class="col"
            >
              Ohar
            </h6>
          </div>
          <hr />
        </div>
      </div>
    </div>
    <div
      class="text-center"
    >
      <a
        href="/main/account/edit"
      >
        <button
          class="btn btn-primary"
        >
          Edit
        </button>
      </a>
    </div>
  </div>
</div>
`;
